core/src/main/scala/org/apache/spark/status/AppStatusListener.scala

```diff
@@ -350,11 +350,16 @@ private[spark] class AppStatusListener(
         val e = it.next()
         if (job.stageIds.contains(e.getKey()._1)) {
           val stage = e.getValue()
-          stage.status = v1.StageStatus.SKIPPED
-          job.skippedStages += stage.info.stageId
-          job.skippedTasks += stage.info.numTasks
-          it.remove()
-          update(stage, now)
+          // Only update the stage if it has not finished already
+          if (v1.StageStatus.ACTIVE.equals(stage.status) ||
+              v1.StageStatus.PENDING.equals(stage.status)) {
+            stage.status = v1.StageStatus.SKIPPED
+            job.skippedStages += stage.info.stageId
+            job.skippedTasks += stage.info.numTasks
+            job.activeStages -= 1
+            it.remove()
```

Comment:
removal from iterator should always happen

Reply:
For already completed stages, we will leave the removal of stage to happen in either onTaskEnd or onStageCompleted event. This ensures that stage metrics are updated even when onJobEnd event is received before onTaskEnd event.

Reply:
I am not sure I follow - if that is the case, why are we doing this for active stages here ? onStageCompleted/onTaskEnd would be fired for active stages as well.

Reply:
Btw, there is an existing bug that we are not updating pool, etc which we do in onStageCompleted ...

Reply:
I think the assumption here is that we will always receive onStageCompleted event before onJobEvent. If that does not occur for some reason, then any active stages are marked as skipped.
I don't know the scenario when onStageCompleted event is not received before onJobEnd event (or received at all). Let me look further into it. Additionally, I will also fix the bug for updating pool.

Reply:
I looked further into it and I could not find any case where onJobEnd is received before onStageCompleted. So the code that iterates over active stages in onJobEnd most likely handles some unforeseen edge scenario.

I have fixed the bug with updating pool though.

Reply:
This can happen when events get dropped ...
Spark makes best case effort to deliver events in order; but when events get dropped, UI becomes inconsistent. I assume this might be an effort to recover in that case ? Any thoughts @tgravescs, @vanzin ?

Reply:
In your question, this == this PR?

If so, no, that's not what it's fixing. Task end events can "naturally" arrive after the stage end event in the case of a stage failure, and this code was missing that case.

When event drops occur, a lot of things get out of sync, and this change wouldn't fix that. It perhaps could make it a little worse: if a task end event does not arrive, then maybe with this change the stage will never be actually removed from the live stages map. Not sure how easy it would be to recover from that though, since dropped events could probably cause other sorts of leaks in this class too, but I also feel that's a separate issue.

(Also, hopefully, dropped events for this listener should be less common in 2.3 after the listener bus changes.)

Reply:
To clarify, I was referring to 'this' being job end event received before stage end (for a stage which is part of a job).

I was not referring to task end event's (those can come in after stage or job end's).

Thanks for clarifying @vanzin ... given the snippet is not trying to recover from events drop, wondering why "non"-skipped stages would even be in the list : I would expect all of them to be skipped ?

Reply:
They would be in the list if the task end event arrives late, right? (Haven't really re-read the code to be sure.)

Unless it's guaranteed that the task end event will arrive before the job end event, unlike the case for the stage end one.

Reply:
This code now only handles the scenario when onStageCompleted event is dropped (not received). If we don't want to handle that scenario, then we can remove this part of the code altogether.